site/content/blog/2022-03-22-schema-migrations.md

```diff
@@ -0,0 +1,126 @@
+
+So far, it seems like using a schema migration tool seems a good way of managing this matter. Here [morph](https://github.com/mattermost/morph/) comes to the stage. It’s a schema migration tool that is built to help improve Mattermost schema migrations.
+
+## Morph
+
+Morph is a schema migration tool that could be used programmatically with a library and from a CLI. It stores the schema version in a table where all applied migrations are persisted.
+It’s a flexible library that can be used as a driver interface implementation, a source to read migration scripts from. The engine uses a dependency injection pattern so that any implementation could be used with the library. And the interfaces of the dependencies are kept minimal to avoid confusion.
+
+The morph engine can be initialized as following:
+
+```Go
+engine, err := morph.New(context.Background(), driver, src)
+if err != nil {
+    return err
+}
+defer engine.Close()
+
+err = engine.ApplyAll()
+```
+
+It’s worth mentioning that MySQL, PostgreSQL, and SQLite implementations are provided within the project. The interface is defined as follows:
+
+```Go
+type Driver interface {
+    Ping() error 
+    Close() error
+    Apply(migration *models.Migration, saveVersion bool) error
+    AppliedMigrations() ([]*models.Migration, error)
+    SetConfig(key string, value interface{}) error
+}
+```
+
+Morph engine expects the migration scripts to be parsed and accessible from an interface: `source.Source` which is a minimal interface as:
+
+```Go
+type Source interface {
+    Migrations() (migrations []*models.Migration)
+}
+```
+
+The `models.Migration` is a simple struct that contains migration name, version, direction, and the migration statements as an `io.ReadCloser` interface can be a file essentially. The `models.Migration` struct can be initialized from the files and the built-in file source expects files to have a specific format.
+
+The migrations files should have up and down versions. The tool expects each migration to be reversible, and the naming of the migration should be in the following form:
+
+```none
+{order_number}_{migration_name}.{(up|down).sql}
+```
+
+The first part will be used to determine the order in which the migrations should be applied and the next part until the `up|down.sql` suffix will be the migration name.
+
+For example:
+
+```none
+0000000001_create_user.up.sql
+0000000001_create_user.down.sql
+```
+
+Apart from the essential arguments, morph can be initialized with a locking option which enables a lock mechanism (to mimic advisory locking) that’s been builtin with the tool. It is being used for cases where multiple applications could execute the migrations at the same time. If the database driver implements the following interface, morph can be used safely in multi-application deployments:
```

Comment:
We should say why we have stayed away from using advisory locks and used our own locking mechanism.